### PR TITLE
Restructure for startup, readiness, liveness and shutdown

### PR DESCRIPTION
**Pull request by seabaylea**
This carries forward the work from @micksi to resolve:
CloudNativeJS/cloud-health-connect#5

as well as implements his suggestions to separate out startup, readiness, liveness and shutdown.

This will form the basis of v2 of cloud-health.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,41 +22,48 @@ for use with Kubernetes and Cloud Foundry based clouds.
...
+ 
+4. On a `SIGTERM` signal for shutdown  
+ Promises that are created as part of a `ShutdownCheck` and registered using `registerShutdownCheck` are executed when the process receives a `SIGTERM` making it possible to clean up any resources used by the application. If the shutdown promises are still running, calls to `getReadinessStatus()`, `getLivenessStatus()` and `getStatus()` return `STOPPING`. Once the promises complete, `STOPPED` is reported.
+
+#### Readiness vs. Liveness
 
+Liveness and readiness checks are executed in the same way but are executed independently (based on calls to `getLivenessStatus()` or `getReadinessStatus()`) or together (based on calls to `getStatus()`).
 
+The difference between liveness and readiness is intended to be purpose: readiness should be used to denote whether an application is "ready" to receive requests, and liveness should be used to denote whether an application is "live" (vs. in a state where it should be restarted.
```
Comment (tomleah) [inline README.md]: "(" in last sentence to be removed

File: src/healthcheck/HealthChecker.ts [hunk 2]
```diff
@@ -86,20 +88,27 @@ class HealthChecker {
...
       this.shutdownRequested = true;
-      for (let plugin in this.shutdownPlugins) {
-        if (this.shutdownPlugins[plugin] != undefined) {
-          this.shutdownPlugins[plugin].runCheck()
-        }
-      }
+      this.shutdownPlugins.map((check) => {
+        check.runCheck()
+      })
+      //for (let plugin in this.shutdownPlugins) {
```
Comment (tomleah) [inline src/healthcheck/HealthChecker.ts]: nit: we can remove this commented code as the above works

**Review by tomleah — COMMENTED** [review]
Have started reviewing.. documentation looks good

**Comment by tomleah** [comment]
Would appreciate a review from @micksi or another contributor/maintainer in addition to mine

**Comment by mhamwala** [comment]
I have also started reviewing

File: README.md (hunk 1 above)
Comment (seabaylea) [reply README.md]: Bracket added.

File: src/healthcheck/HealthChecker.ts (hunk 2 above)
Comment (seabaylea) [reply src/healthcheck/HealthChecker.ts]: Removed.

File: README.md [hunk 3]
```diff
@@ -21,41 +22,48 @@ for use with Kubernetes and Cloud Foundry based clouds.
 
 ## Using Cloud Health
 
-Cloud Health allows you to register promises which are executed during the three phases of your application, and allows you to call `getStatus()` to return a promise which resolves to whether the application is `STARTING`, `UP`, `DOWN`, `STOPPING` or `STOPPED`.
+Cloud Health allows you to register promises which are executed during the three phases of your application, and allows you to call `getLivenessStatus()`, `getReadinessStatus()`, or a combined `getStatus()` to return a promise which resolves to whether the application is `STARTING`, `UP`, `DOWN`, `STOPPING` or `STOPPED`.
 
-1. At startup for "readiness"  
- Promises that are created as part of a `ReadinessCheck` and registered using `registerReadinessCheck` are executed at startup and can be used to execute any code that must complete before your application is ready. If the startup promises are still running, calls to `getStatus()` return `STARTING`. Once the promises complete, `DOWN` is reported if there were any failures, or the "liveness" promises are then executed.
+1. At startup for "startup"
```
Comment (mhamwala) [inline README.md]: should this be: At runtime?

File: README.md [hunk 4]
```diff
@@ -21,41 +22,48 @@ for use with Kubernetes and Cloud Foundry based clouds.
...
 
-1. At startup for "readiness"  
- Promises that are created as part of a `ReadinessCheck` and registered using `registerReadinessCheck` are executed at startup and can be used to execute any code that must complete before your application is ready. If the startup promises are still running, calls to `getStatus()` return `STARTING`. Once the promises complete, `DOWN` is reported if there were any failures, or the "liveness" promises are then executed.
+1. At startup for "startup"  
+ Promises that are created as part of a `StartupCheck` and registered using `registerStartupCheck` are executed at startup and can be used to execute any code that must complete before your application is ready. If the startup promises are still running, calls to `getLivenessStatus()`, `getReadinessStatus()`, and `getStatus()`,  return `STARTING`. Once the promises complete, `DOWN` is reported if there were any failures, or the "liveness" promises are then executed.
   
 2. At runtimes for "liveness"
```
Comment (mhamwala) [inline README.md]: At runtime*

**Comment by mhamwala** [comment]
we could possibly add the `npm install @cloudnative/health` cmd to the readme

File: README.md (hunk 4 above)
Comment (seabaylea) [reply README.md]: Done

**Comment by seabaylea** [comment]
@mhamwala I've just added an install step.

File: README.md [hunk 5]
```diff
@@ -43,12 +43,19 @@ Liveness and readiness checks are executed in the same way but are executed inde
 The difference between liveness and readiness is intended to be purpose: readiness should be used to denote whether an application is "ready" to receive requests, and liveness should be used to denote whether an application is "live" (vs. in a state where it should be restarted).   
 
 ### Using Cloud Health with Node.js
-1. Set up a HealthChecker:
+
+1. Installation:
+
```
Comment (mhamwala) [inline README.md]: nit: spacing around this code block causes it to indent, further than the other steps

Comment (seabaylea) [reply README.md]: Done

**Review by tomleah — APPROVED** [review]
Tests look good, rubber stamp LGTM

**Review by mhamwala — APPROVED** [review]
rubber stamp LGTM.

**Pull request merged**